branchpro/models.py

```diff
@@ -97,12 +102,71 @@ def __init__(self, initial_r, serial_interval):
 
         # Invert order of serial intervals for ease in _normalised_daily_mean
         self._serial_interval = np.asarray(serial_interval)[::-1]
-        self._initial_r = initial_r
+        self._r_profile = np.array([initial_r])
         self._normalizing_const = np.sum(self._serial_interval)
 
+    def set_r_profile(self, new_rs, start_times, last_time=None):
+        """
+        Creates a new R_t profile for the model.
+
+        Parameters
+        ----------
+        new_rs: sequence of new time-dependent values of the reproduction
+            numbers.
+        start_times: sequence of the first time unit when the corresponding
+            indexed value of R_t in new_rs is used. Must be an ordered sequence
+             and without duplicates or negative values.
+        last_time: total evaluation time; optional.
+
+        """
+        # Raise error if not correct dimensionality of inputs
+        if np.asarray(new_rs).ndim != 1:
+            raise ValueError(
+                'New reproduction numbers storage format must be 1-dimensional'
+                )
+        if np.asarray(start_times).ndim != 1:
+            raise ValueError(
+                'Starting times values storage format must be 1-dimensional')
+
+        # Raise error if inputs do not have same shape
+        if np.asarray(new_rs).shape != np.asarray(start_times).shape:
+            raise ValueError('Both inputs should have same number of elements')
+
+        # Raise error if start times are not non-negative and increasing
+        if np.any(np.asarray(start_times) < 0):
+            raise ValueError('Times can not be negative.')
+        if np.any(np.asarray(start_times)[:-1] >= np.asarray(start_times)[1:]):
+            raise ValueError('Times must be increasing.')
+
+        # Ceil times to integer numbers a
+        times = np.ceil(start_times).astype(int)
+
+        # Create r profile
+        r_profile = []
+
+        # Fill in initial r from day 1 up to start time
+        initial_r = self._r_profile[0]
+        r_profile += [initial_r] * (times[0] - 1)
+
+        # Fill in later r's
+        time_intervals = times[1:] - times[:-1]
+        for time_id, time_interval in enumerate(time_intervals):
+            # Append r for each time unit
+            r_profile += [new_rs[time_id]] * time_interval
+
+        # Add final r
+        if last_time:
+            final_interval = last_time - start_times[-1] + 1
+            r_profile += [new_rs[-1]] * final_interval
+        else:
+            r_profile += [new_rs[-1]]
```

Comment:
I wouldn't be hugely surprised if there's a faster way to generate this r_profile. Let's not change it for now but keep it in mind in case it becomes a bottleneck